src/ast/nodes/index.ts

```diff
@@ -111,12 +112,12 @@ const nodes: {
 	ThrowStatement,
 	TryStatement: StatementBase,
 	UnaryExpression,
+	UnknownNode,
 	UpdateExpression,
 	VariableDeclarator,
 	VariableDeclaration,
 	WhileStatement,
 	YieldExpression
 };
 
-export { NodeType } from './NodeType';
 export default nodes;
```

Comment:
If you change this to `export { nodes as default }` that should avoid the circular dependency issue by ensuring a live binding.

Reply:
I just tried it but it does not seem to work. As the circular dependency is still present (though with the live binding), the modules are executed in an order where the actual node definitions, e.g. `ArrayExpression`, are evaluated before the base class `NodeBase` is evaluated. As TypeScript then wants to synchronously access `NodeBase.prototype`, rollup crashes. But maybe there is another way without passing it explicitly as a parameter.

Reply:
Which is the first access to fail?

Reply:
It is the `NodeBase.prototype` access in `ArrayExpression`:)

In order to improve encapsulation, I currently investigate if instead of attaching the module to each node, I attach a module proxy object with all the interface that the nodes need. That way we could reduce the API surface of Module and Graph but I could also just attach the node constructors there to avoid the circularity.

Reply:
It should be possible to rearrange the import statements to ensure circular execution happens in the desired order.

I just checked master and we currently have `shared/Node.ts` executing before `ArrayExpression.ts`, so perhaps somehow this post-order changed in this PR? If so, it should be possible to ensure `shared/Node.ts` comes first in the post-order to maintain the correct circular execution order.

Reply:
It might be but I'm not a big fan. For once, I have to add an exception to ignore circular dependency warnings when building rollup (not a big issue but still). My bigger concern, however, is that new code can always break the execution order again. My new approach, which also improves encapsulation, is to create a special wrapper layer between the module and the AST/scopes/variables that also contains the node constructors. That way, ast nodes never directly access module or graph which means those can be easily refactored as long as the wrapper layer is refactored accordingly. I could even make a few more fields private on `Module`.

As it turns out, instead of making things slower this wrapper layer seems to make things slightly faster, probably because all wrappers have the same shape and thus property access can be monomorphic.